### PR TITLE
SAK-15574 Assignments: automatic late penalty (flat or per-day) with per-submission waiver

### PR DESCRIPTION
**Pull request by spencer-dollahite**
https://sakaiproject.atlassian.net/browse/SAK-15574

> **First pass.** This implements the January 2026 T&L/UX mockups with working backend logic so the feature can actually be played with; I expect to revisit details based on testing and feedback. The exact semantics I chose (points-only, whole-day rounding, due-date baseline, floor at zero) are spelled out in [my ticket comment](https://sakaiproject.atlassian.net/browse/SAK-15574?focusedCommentId=255416) awaiting T&L confirmation — happy to adjust any of them.

## Problem

Instructors who apply a late policy ("−5 for late work", "−2 per day late") have to compute and enter every adjusted grade by hand, and the gradebook keeps no trace of the policy. This has been requested since 2009; the T&L/UX group revisited it this January, confirmed it is still wanted, and posted mockups for both the assignment settings and the grader (attached to the ticket).

## Feature

**Settings.** For Points-scale assignments only, the Grading section gains a "Late Penalty" option with two modes per the mockup: deduct a flat number of points for late submissions, or deduct N points per day late. Stored as assignment properties (`late_penalty_type`, `late_penalty_value`) — no schema change. The option is validated (positive point value, grade-style precision) and cleared if the grade type moves away from Points; it is not offered together with peer assessment in this pass.

**Semantics.** A submission is late when it was submitted after the due date, matching the existing LATE flag everywhere else. Any part of a day counts as a full day (one minute late = 1 day, 24h01m = 2 days). The effective grade is floored at zero.

**Non-destructive by design.** The instructor grades the work on its merits and the raw grade is what's stored — the penalty is computed wherever the grade is read or pushed, in `AssignmentService.getGradeForSubmitter` (so every student-facing surface inherits it) and in the gradebook integration (so the released score matches what the student sees). A new pure `LatePenaltyCalculator` in assignment-api holds the arithmetic, in the same scaled-integer space grades are stored in. Lessons completion conditions compare the same penalized value.

**Grader.** Per the mockup, the grader shows the raw grade in the input, a red "−X late" annotation next to the max-points label, and a per-submission **Ignore Late Penalty** checkbox (stored as a submission property) that waives the penalty for that student. Because grading inputs must never round-trip a computed value, everything that prefills an editable grade field (grader DTOs, legacy per-submitter override state) is explicitly fed the raw grade — otherwise a penalized grade saved back would compound on every save.

## Update (post-review): intervals and percentages

Following T&L discussion on the ticket, the penalty model is generalized: deduct **once, per hour, per day, or per week** late (any partial period counts as a full period), in **fixed points or a percentage of the max grade** per application (the Canvas model, e.g. "10% per day"). An absent unit property means points, so nothing saved earlier changes behavior; a percentage is validated to at most 100. The settings block now uses interval radios with a single amount + unit row. Covered by the expanded calculator test matrix (24 tests).

## Known limitations (intentional for a first pass)

- **Due-date changes after release**: the penalty is computed on read, so displays update immediately, but a score already pushed to the gradebook stays as-is until the submission is next saved/released. A bulk re-push on assignment save would fix this but is a larger change.
- **Baseline is the due date**, even when a resubmission window or extension pushes the accept-until later — consistent with the existing LATE flag, and the Ignore checkbox is the per-student escape hatch. If T&L prefers extensions to move the baseline, that's a contained change in the calculator.
- **Legacy (non-grader) grading UI** shows and edits raw grades and gets no Ignore checkbox in this pass; since the penalty is applied server-side on read/push, values remain correct either way.
- **Peer-assessment assignments** are excluded (the peer flow overwrites the stored grade with the computed average; stacking a penalty on top needs its own discussion).
- **Tests & Quizzes** equivalent is SAK-52267 and stays a separate follow-up.

## Testing

- `LatePenaltyCalculatorTest` (18 tests): flat and per-day math, partial-day ceiling, exactly-24h boundary, floor at zero, on-time/at-due-date not penalized, ignore flag, null dates, ungraded passthrough, malformed/zero/negative config inert, scale factors 10 and 100, decimal penalty values.
- `AssignmentServiceTest` (+4): penalized display grade end-to-end through `getGradeForSubmitter`, ignore-flag waiver, stale penalty config inert on non-Points assignments, group assignments (per-submitter override grades penalized too).
- `AssignmentToolUtilsLatePenaltyTest` (2): the gradebook push sends the effective grade while the stored submission grade stays raw; the ignore flag round-trips through `gradeSubmission` options.
- sakai-grader web tests (+2): annotation and checkbox render for a penalized late submission, checkbox hides the annotation and flows into the save form data; no controls when no penalty applies.
- Manually verified on a local instance driving the real UI: created a 100-point assignment with a 2/day penalty, student submitted ~30 minutes late, grader showed "−2.00 late", graded 90 and released → student view and Gradebook both show 88 while the stored grade remains the raw 90; checking Ignore Late Penalty and re-releasing returns both to 90. Settings survive the edit round-trip and only appear for the Points scale.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added automatic late-penalty support for points-based assignments, including flat and per-hour/day/week deductions, configurable as points or % of max.
  * Added per-submission “ignore late penalty” option with an instructor UI control and grader late-penalty annotation.
  * Added grade displays to support viewing both raw (no penalty) and effective (with penalty) values.

* **Bug Fixes**
  * Late-penalty behavior now stays consistent across grading views and gradebook integration, without altering stored raw grades.
  * Penalties are correctly suppressed when not applicable or when the submission is flagged to ignore.

* **Tests**
  * Added/expanded unit and UI integration tests for penalty math, edge cases, and option persistence.

### CODE REVIEW

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 1]
```diff
@@ -8629,6 +8642,19 @@ private void setNewAssignmentParameters(RunData data, boolean validify) {
...
+            Integer scaleFactor = assignmentService.getScaleFactor();
+            try {
+                if (StringUtils.isNotEmpty(assignmentRef)) {
+                    Assignment assignment = assignmentService.getAssignment(assignmentId);
+                    if (assignment != null && assignment.getScaleFactor() != null) {
+                        scaleFactor = assignment.getScaleFactor();
+                    }
+                }
+            } catch (IdUnusedException | PermissionException e) {
+                log.error(e.getMessage());
```
Comment (ern) [inline assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: uninformative log message

```suggestion
                log.warn("Could not fetch assignment with id [{}], {}", assignmentId, e.toString());
```

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 2]
```diff
@@ -8629,6 +8642,19 @@ private void setNewAssignmentParameters(RunData data, boolean validify) {
         String option = params.get("option");
 
         if (!"attach".equals(option)) {
+            // an existing assignment's own scale factor governs grade and penalty precision
+            Integer scaleFactor = assignmentService.getScaleFactor();
+            try {
+                if (StringUtils.isNotEmpty(assignmentRef)) {
```
Comment (ern) [inline assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: logic issue, checks if assignmentRef is not empty but then never uses it?

File: assignment/api/src/java/org/sakaiproject/assignment/api/AssignmentConstants.java [hunk 3]
```diff
@@ -229,6 +229,29 @@ public final class AssignmentConstants {
      */
     public static final String ALLOW_RESUBMIT_CLOSETIME = "allow_resubmit_closeTime";
     public static final String ALLOW_EXTENSION_CLOSETIME = "allow_extension_closeTime"; //constant for Extension, very similar to Resubmission
+    /**
+     * SAK-15574 assignment property holding the late penalty mode, a
```
Comment (ern) [inline assignment/api/src/java/org/sakaiproject/assignment/api/AssignmentConstants.java]: avoid SAK references in comments, it just adds noise

File: assignment/api/src/java/org/sakaiproject/assignment/api/LatePenaltyCalculator.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                return valueOf(value.trim().toUpperCase());
+            } catch (IllegalArgumentException e) {
+                log.debug("Unrecognized late penalty unit [{}], treating as POINTS", value);
+                return POINTS;
+            }
+        }
+    }
+
+    private LatePenaltyCalculator() {
+    }
```
Comment (ern) [inline assignment/api/src/java/org/sakaiproject/assignment/api/LatePenaltyCalculator.java:102]: if you want to prevent instantiation maybe an exception should be thrown here? makes it very clear that this is a static only class

File: assignment/impl/src/test/org/sakaiproject/assignment/impl/AssignmentServiceTest.java [hunk 5]
```diff
@@ -2032,6 +2032,105 @@ public void gradeDisplay() {
         Assert.assertEquals("self", assignmentService.getGradeDisplay("self", Assignment.GradeType.GRADE_TYPE_NONE, null));
     }
 
+    @Test
```
Comment (ern) [inline assignment/impl/src/test/org/sakaiproject/assignment/impl/AssignmentServiceTest.java:2035]: good tests

File: assignment/tool/src/java/org/sakaiproject/assignment/entityproviders/AssignmentEntityProvider.java [hunk 6]
```diff
@@ -1655,6 +1662,32 @@ public List<SimpleSubmission> getItemSubmission(EntityView view, Map<String, Obj
...
 
+    /**
+     * SAK-15574 the display grade for a submitter WITHOUT any late penalty
+     * applied. Grading inputs must prefill with this raw value so the
+     * penalized effective grade never round-trips through a save and
+     * compounds. The group override resolution here intentionally mirrors
+     * {@code AssignmentServiceImpl#getGradeForSubmitter} minus the penalty
+     * step — keep the two in sync.
+     */
+    private String rawGradeDisplay(AssignmentSubmission submission, String submitter) {
```
Comment (ern) [inline assignment/tool/src/java/org/sakaiproject/assignment/entityproviders/AssignmentEntityProvider.java:1670]: pretty sure there is already a service method which provides this info

File: assignment/tool/src/test/org/sakaiproject/assignment/tool/AssignmentToolUtilsLatePenaltyTest.java [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+ * (late-penalized) grade while the stored submission grade stays raw, and the
+ * per-submission ignore flag must round-trip through gradeSubmission options.
+ */
+@RunWith(MockitoJUnitRunner.Silent.class)
+public class AssignmentToolUtilsLatePenaltyTest {
+
+    private static final String SITE_ID = "site1";
+    private static final String STUDENT = "student1";
+
+    @Mock private AssignmentService assignmentService;
```
Comment (ern) [inline assignment/tool/src/test/org/sakaiproject/assignment/tool/AssignmentToolUtilsLatePenaltyTest.java:65]: probably not ideal to mock the AssignmentService when running test in assignments, this is generally a smell (Mocking an internal service).

File: assignment/tool/src/test/org/sakaiproject/assignment/tool/AssignmentToolUtilsLatePenaltyTest.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+        String assignmentRef = AssignmentReferenceReckoner.reckoner().assignment(assignment).reckon().getReference();
+        String submissionRef = AssignmentReferenceReckoner.reckoner().submission(submission).reckon().getReference();
+        String associate = "gbitem1";
+
+        when(assignmentService.getToolId()).thenReturn("sakai.assignment.grades");
+        when(assignmentService.getAssignment("a1")).thenReturn(assignment);
+        when(assignmentService.getSubmission("s1")).thenReturn(submission);
+        // the service computes 90.00 - 5 = 85.00 for this late submission
+        when(assignmentService.getGradeForSubmitter(submission, STUDENT)).thenReturn("85.00");
```
Comment (ern) [inline assignment/tool/src/test/org/sakaiproject/assignment/tool/AssignmentToolUtilsLatePenaltyTest.java]: this mocking logic is what should not be mocked when testing internal services.

AI typically gets this wrong!

**Comment by ern** [comment]
@spencer-dollahite thanks for the PR much appreciated, couple of tips when using AI

- open the PR as a draft initially this signals that your not yet ready for human review, especially if your opening a burst of PR's
- when using AI it is expected that you have tested the PR yourself and it is working as you expect
- you shouldn't rely on community resources to do review and testing until you have performed your own testing

AI suggested these tips for you!
- Audit Automated Unit Tests & Mocking: AI models love to pass tests by simply mocking out the very logic they are supposed to be testing. Always review generated tests to ensure they are executing the real code pathways and asserting actual behavior, rather than over-mocking internal services.

- Check for Code Duplication and Native Patterns: AI often writes isolated, self-contained code that works but misses existing project-wide utilities, enums, or constants. Ensure the AI didn't recreate logic or hardcode strings that already exist elsewhere in the codebase.

- Sanity Check Boundary Conditions & Data Types: AI is notoriously prone to subtle edge-case errors, such as missing null/empty checks, assuming the wrong scale/precision for numeric inputs, or misconfiguring HTML form validation attributes (e.g., using a plain text input instead of a constrained numeric input).

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java (hunk 1 above)
Comment (spencer-dollahite) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: Good call — switched to `log.warn` with the assignment id + exception (your suggestion). 3db5f45

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java (hunk 2 above)
Comment (spencer-dollahite) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: Yeah, that was inconsistent. The guard now checks `assignmentId` (the value actually used for the lookup); `assignmentRef` only sticks around as the raw param it's derived from. 3db5f45

File: assignment/api/src/java/org/sakaiproject/assignment/api/AssignmentConstants.java (hunk 3 above)
Comment (spencer-dollahite) [reply assignment/api/src/java/org/sakaiproject/assignment/api/AssignmentConstants.java]: Agreed, dropped the noise. I pulled the `SAK-#####` prefixes out of comments/Javadoc across the whole PR — not just here — including the JS, the Velocity template, and the bundle comments. 3db5f45

File: assignment/api/src/java/org/sakaiproject/assignment/api/LatePenaltyCalculator.java (hunk 4 above)
Comment (spencer-dollahite) [reply assignment/api/src/java/org/sakaiproject/assignment/api/LatePenaltyCalculator.java:102]: Done — the private constructor now throws an `AssertionError` so the static-only intent is explicit. 3db5f45

File: assignment/impl/src/test/org/sakaiproject/assignment/impl/AssignmentServiceTest.java (hunk 5 above)
Comment (spencer-dollahite) [reply assignment/impl/src/test/org/sakaiproject/assignment/impl/AssignmentServiceTest.java:2035]: Thanks! 🙏

File: assignment/tool/src/java/org/sakaiproject/assignment/entityproviders/AssignmentEntityProvider.java (hunk 6 above)
Comment (spencer-dollahite) [reply assignment/tool/src/java/org/sakaiproject/assignment/entityproviders/AssignmentEntityProvider.java:1670]: You're right, there was one hiding in plain sight. I added `AssignmentService#getRawGradeForSubmitter` backed by a shared private resolver that `getGradeForSubmitter` also calls (the two paths only differ by whether the late penalty is applied), and this provider now just delegates to it — no more hand-rolled twin to keep in sync. 3db5f45

File: assignment/tool/src/test/org/sakaiproject/assignment/tool/AssignmentToolUtilsLatePenaltyTest.java (hunk 7 above)
Comment (spencer-dollahite) [reply assignment/tool/src/test/org/sakaiproject/assignment/tool/AssignmentToolUtilsLatePenaltyTest.java:65]: Fair point, and you're right that I got this wrong. I removed the test that mocked `getGradeForSubmitter` to return the penalized value — it was just asserting its own mock. The penalty math now runs against the real `AssignmentServiceImpl` + `LatePenaltyCalculator` in the impl module's `AssignmentServiceTest`. The only mock left in this class is the persistence seam (`updateSubmission`), not the computation. 3db5f45

File: assignment/tool/src/test/org/sakaiproject/assignment/tool/AssignmentToolUtilsLatePenaltyTest.java (hunk 8 above)
Comment (spencer-dollahite) [reply assignment/tool/src/test/org/sakaiproject/assignment/tool/AssignmentToolUtilsLatePenaltyTest.java]: Removed — that stub is exactly what I took out. Real-code coverage lives in `AssignmentServiceTest` now (late/flat/on-time/ignore-flag/non-points/group-override), so nothing here fakes the grade anymore. Thanks for the nudge. 3db5f45

**Comment by spencer-dollahite** [comment]
Thanks for the thorough review and the tips, @ern — genuinely appreciated, and point taken on all of it. I've pushed 3db5f45 addressing each inline comment: reused the service for the raw grade instead of the hand-rolled twin, dropped the ticket refs from comments, made the utility constructor throw, and reworked the late-penalty test so the math runs against the real service + calculator rather than a mock (that one was a good catch — exactly the over-mocking trap you called out).

On process: fair on all three. I'll open future work as draft until I've done my own testing pass, and lean on my own environment for verification before asking for eyes. Thanks again for putting up with the back-and-forth.
